### DISCUSSION TITLE
BlockGap exception for blocks in the top level container

### DISCUSSION THREAD

**Discussion by crs1138** [WordPress/gutenberg#46902] [Q&A]
Our design assumes that the blockGap for all block in the main container of the content on the page is 32px. Is it possible to create an exception for the block `core/paragraph` in the context of the main content to be 0px? 

Here is an excerpt from our _theme.json_:
```json
"styles": {
    "spacing": {
      "blockGap": "32px"
    },
    "blocks": {
      "core/paragraph": {
        "spacing": {
          "blockGap": "0px"
        }
      },
      "core/group": {
        "spacing": {
          "blockGap": "0px"
        }
    },
    ...
}
```
The blocks (including `core/paragraph`s) within the `core/group` block has `blockGap: 0px`, but within the main content this does not seem to work.

**Comment by devfle** [comment]
You could build a workaround using the following filter:
https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#blocks-registerblocktype

Just manipulate the settings object.

**Reply by devfle** [reply]
And if you do not want to disable the gap feature for paragraphs you could also manipulate just the block attributes from the paragraph block: https://developer.wordpress.org/block-editor/reference-guides/filters/block-filters/#blocks-getblockattributes

```javascript
addFilter('blocks.getBlockAttributes', 'theme/paragraph/changeGap', (attributes, settings) => {
 if (settings.name !== 'core/paragraph') return attributes;
 ...
 ...
 return attributes;
}, 10)
```

**Reply by crs1138** [reply]
Thanks @devfle I'll tinker with the filters. 

The second suggestion is more in line with our actual use-case as I had to simplify it for this discussion. 

Perhaps, in a future release, it could be possible to assign a blockGap value for individual types of blocks in theme.json. The way I understand it now, it is a value gaps between block within the container the value is assigned to. However, that seems somewhat limiting and heading or paragraph blocks can be a great example of such limitation.

**Reply by devfle** [reply]
If I can help any further, feel free to let me know. 👍

**Reply by crs1138** [reply]
I'm really confused. It does not seem to be possible to change the `blockGap` value of a specific type of blocks. The `blockGap` value is applied to a container like a `core/group` and is applied to all the blocks within that container via the CSS property `margin-block-start` which in our case (`write-mode: ltr`) translates into the top margin. I can change the `core/paragraph` default style in theme.json to have a specific top margin value, however, the `margin-block-start` ends up having priority in this case. I don't understand, how to set that all blocks will have the same blockGap value (for example 32px) and there will be an exception for `core/paragraph` that will be different (for example 16px).

**Reply by crs1138** [reply]
This seems to be describing the same issue – https://github.com/WordPress/gutenberg/issues/35267
Is it still the case then?

**Comment by t-hamano** [comment]
Perhaps this topic would be relevant to #43404.

**Comment by tellthemachines** [comment]
Hi @crs1138 ! The `blockGap` setting only applies to container blocks (e.g. Group, Row, Cover; any block that can contain other blocks). You can however set `margin` on Paragraph blocks, and since #47858 was merged it should now have the intended effect.
